torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -182,12 +218,16 @@ class _OverlapInfo():
     """
     def __init__(self) -> None:
         self.status: _OverlapStatus = _OverlapStatus.UNINITIALIZED
+        self.uniform_bucket_assignment: bool = False
 
         # Modified per bucket reconstruction
         self.params_per_bucket: List[List[torch.Tensor]] = []
         self.params_per_rank: List[List[torch.Tensor]] = \
             [[] for _ in range(dist.get_world_size())]
         self.offsets: Dict[int, int] = {}
+        self.assigned_ranks_per_bucket: List[Set[int]] = []
```

Comment:
I want to discuss the data structure choice for the inner dimension of `assigned_ranks_per_bucket`.

There are two usages:
(1) Given a `rank` check if `rank in assigned_ranks_per_bucket[bucket_index]`. This happens > 1 time per rank per bucket per iteration.
(2) Iterate over `assigned_ranks_per_bucket[bucket_index]` in a deterministic order across all ranks. This happens 1 time per rank per bucket per iteration.

I currently use a `Set[int]`, which means per rank per bucket per iteration, (1) is `O(1)` but (2) is `O(n log n)` where `n` is the length of `assigned_ranks_per_bucket[bucket_index]`. In this case, `n` is never large since we should not be breaking a DDP bucket into many subsets, as that would imply either a large `bucket_cap_mb` or a many small ZeRO bucket assignments, both of which are most likely suboptimal. Moreover, Python's Timsort is pretty performant. As such, we probably do not need to stress over this. However, I still wanted to bring up some other options to be thorough.

We could consider using a `SortedList[int]` (using `sortedcontainers.SortedList`) or a `OrderedDict[int, ...]` (to simulate an `OrderedSet`, which I do not believe is formally provided by Python).

For `SortedList[int]`, are we allowed this dependency on an external library? It would make (1) `O(log n)` (probably using `bisect.bisect()` or something similar) and make (2) zero cost. 
For `OrderedDict[int, ...]`, we would have to use dummy values (e.g. by repeating the key as the value). It would make (1) `O(1)` and make (2) zero cost, but it would require a bit more memory.

Reply:
> For `SortedList[int]`, are we allowed this dependency on an external library?

It does not seem worth it to introduce this external dependency for this purpose, given that n is never large.